src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts

```diff
@@ -0,0 +1,77 @@
+import { UseCaseInterface } from '../UseCaseInterface'
+import { inject, injectable } from 'inversify'
+import TYPES from '../../../Bootstrap/Types'
+import { GetUserFeaturesDto } from './GetUserFeaturesDto'
+import { UserRepositoryInterface } from '../../User/UserRepositoryInterface'
+import { GetUserFeaturesResponse } from './GetUserFeaturesResponse'
+import { Permission, PermissionName, RoleName } from '@standardnotes/auth'
+import { UserSubscription } from '../../User/UserSubscription'
+import { Feature, Features } from '@standardnotes/features'
+import { getSubscriptionNameForRoleName } from '../../Role/RoleToSubscriptionMap'
+
+type PermissionWithExpiresAt = Permission & {
+  expiresAt: number
+}
+
+@injectable()
+export class GetUserFeatures implements UseCaseInterface {
+  constructor(
+    @inject(TYPES.UserRepository) private userRepository: UserRepositoryInterface,
+  ) {
+  }
+
+  async execute(dto: GetUserFeaturesDto): Promise<GetUserFeaturesResponse> {
+    const { userUuid } = dto
+
+    const user = await this.userRepository.findOneByUuid(userUuid)
+
+    if (user === undefined) {
+      return {
+        success: false,
+        error: {
+          message: `User ${userUuid} not found.`,
+        },
+      }
+    }
+
+    const userRoles = await user.roles
+    const userSubscriptions = await user.subscriptions
+    const permissions: PermissionWithExpiresAt[] = []
+
+    await Promise.all(userRoles.map(async (role) => {
+      const subscriptionName = getSubscriptionNameForRoleName(role.name as RoleName)
+      const expiresAt = (userSubscriptions.find(subscription => subscription.planName === subscriptionName) as UserSubscription).endsAt
+
+      const rolePermissions = await role.permissions
+
+      rolePermissions.forEach(rolePermission => {
+        const permissionItem = permissions.find(permission => permission.name === rolePermission.name)
+        /* istanbul ignore else */
+        if (!permissionItem) {
+          permissions.push({
+            ...rolePermission,
+            name: rolePermission.name as PermissionName,
+            expiresAt,
+          })
+        } else if (expiresAt > permissionItem.expiresAt) {
+          permissionItem.expiresAt = expiresAt
+        }
+      })
+    }))
+
+    const features: Feature[] = []
+    permissions.forEach(permission => {
+      const featureItem = Features.find(feature => feature.identifier === permission.name) as Feature
+      features.push({
+        ...featureItem,
+        expiresAt: permission.expiresAt,
+      })
+    })
```

Comment:
Since we only need to return features and not permissions, maybe we could get rid of the permissions array and combine the two parts in one? This way we wouldn't need the extra iteration over the permissions array at the end 🙂 something like this

```javascript
rolePermissions.forEach(rolePermission => {
  const featureItem = features.find(feature => feature.identifier === rolePermission.name)
  const feature = Features.find(feature => feature.identifier === rolePermission.name)
  if (!featureItem) {
    features.push({
       ...feature,
       expiresAt,
    })
  } else if (expiresAt > featureItem.expiresAt) {
    featureItem.expiresAt = expiresAt
  }
})
```

Perhaps we could name the features array "userFeatures" or something similar so it's less confusing

Reply:
Agreed, done)